### ISSUE TITLE
Try catch blocks in OpenMP

### ISSUE THREAD

**Issue by Craigacp** [dmlc/xgboost#1573]
I'm looking at porting XGBoost to Solaris using the Oracle Studio compilers. I have a version that works on x86_64, but I had to comment out lines 264 and 295 from c_api.cc in XGDMatrixCreateFromCSC, (the API_BEGIN and API_END blocks which add try catch statements). The Studio compiler says that the OpenMP spec disallows try catch blocks around OpenMP blocks (http://www.openmp.org/mp-documents/spec30.pdf page 35) as all exceptions have to be handled by the thread that threw them (so the try catch needs to be inside the pragma).

GCC gets around this problem by adding an implicit try catch inside the pragma (http://stackoverflow.com/questions/13663231/throwing-a-c-exception-inside-an-omp-critical-section) but this means it can't ever throw that exception out of the OpenMP section.

Are the API_BEGIN and API_END calls necessary here, or does the logic need further refactoring?

**Comment by tqchen** [comment]
Indeed that try catch blocks cannot directly catch the exceptions thrown out from OpenMP.

The API_BEGIN and API_END macro mainly aims to provide a general regime for error reporting when something failed elsewhere(not in openmp block). So it is never intentionally to catch something out from OpenMP block.

If that is something compiler hates and does not allow compilation, let us see if we can get around by putting openmp block in an independent function. You are more than welcomed to open a PR on this

**Comment by Craigacp** [comment]
Sure. The modifications to make it all compile also require a small change in DMLC-core (adding defined(__sun) to line 127 of dmlc/base.h and a couple of changes to dmlc-core and rabit's Makefiles so they don't assume gcc (as Studio has different optimisation arguments and library names). I'll put the appropriate strapping around them so they are only turned on on Solaris. I also ran a quick test and it seems to pass the Java unit tests on Solaris/SPARC, but I'll need to do a little more verification to make sure everything is still working.

What's the protocol for getting a change into the dmlc-core and rabit projects so xgboost tracks the appropriate commits?

BTW I'm still waiting on approval from my management before I can create the pull requests for either this or the Windows xgboost4j fixes, so it'll be a while before I actually make them.

**Comment by tqchen** [comment]
Please PR to the dmlc-core and rabit repo respectively, after the changes get merged, we can propose another PR to update the submodules
